nixos/modules/services/misc/soft-serve.nix

```
@@ -0,0 +1,140 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.soft-serve;
+  configFile = format.generate "config.yaml" cfg.settings;
+  format = pkgs.formats.yaml { };
+  docUrl = "https://charm.sh/blog/self-hosted-soft-serve/";
+in
+{
+  options = {
+    services.soft-serve = {
+      enable = mkEnableOption "Enable soft-serve service";
+
+      package = mkPackageOption pkgs "soft-serve" { };
+
+      stateDir = mkOption {
+        type = types.path;
+        default = "/var/lib/soft-serve";
+        description = mdDoc ''
+          The absolute path to the data directory.
+
+          See <${docUrl}>.
+        '';
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "soft-serve";
+        description = mdDoc "User account under which soft-serve runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "soft-serve";
+        description = mdDoc "Group account under which soft-serve runs.";
+      };
+
+      settings = mkOption {
+        type = format.type;
+        default = { };
+        description = mdDoc ''
+          The contents of the configuration file.
+
+          See <${docUrl}>.
+        '';
+        example = literalExpression ''
+          {
+            name = "dadada's repos";
+            log_format = "text";
+            ssh = {
+              listen_addr = ":23231";
+              public_url = "ssh://localhost:23231";
+              max_timeout = 30;
+              idle_timeout = 120;
+            };
+            stats.listen_addr = ":23233";
+            initial_admin_keys = config.dadada.admin.users.dadada.keys;
+          }
+        '';
+      };
+    };
+  };
+
+  config = let stateDir = cfg.stateDir; in mkIf cfg.enable {
+    users.users = mkIf (cfg.user == "soft-serve") {
+      soft-serve = {
+        description = "soft-serve service";
+        home = stateDir;
+        useDefaultShell = true;
+        group = cfg.group;
+        isSystemUser = true;
+      };
+    };
+
+    users.groups = mkIf (cfg.group == "soft-serve") {
+      soft-serve = { };
+    };
+
+    systemd.tmpfiles.rules = [
+      "d '${stateDir}' 0750 ${cfg.user} ${cfg.group} - -"
+      "L+ ${stateDir}/config.yaml - - - - ${configFile}"
+    ];
+
+    systemd.services.soft-serve = {
+      description = "Soft Serve git server 🍦";
```

```suggestion
      description = "Soft Serve git server";
```
descriptions and the like should be emoji free otherwise it might show placeholder characters on some terminals